src/server/gc_worker/gc_worker.rs

```diff
@@ -235,23 +216,52 @@ impl<E: Engine> GcRunner<E> {
                 break;
             }
 
-            let mut txn = MvccTxn::for_scan(
-                self.engine.snapshot_on_kv_engine(start_key, end_key)?,
-                Some(ScanMode::Forward),
-                TimeStamp::zero(),
-                false,
-            );
-            for key in keys {
-                if let Err(e) = self.gc_key(safe_point, &key, &mut txn) {
-                    error!("gc fail"; "key" => %key, "err" => ?e);
+            fn new_txn<S: Snapshot>(snap: S) -> MvccTxn<S> {
+                MvccTxn::for_scan(snap, Some(ScanMode::Forward), TimeStamp::zero(), false)
+            }
+
+            fn flush_txn<E: Engine>(
+                txn: MvccTxn<E::Snap>,
+                limiter: &Limiter,
+                engine: &E,
+            ) -> Result<()> {
+                let write_size = txn.write_size();
+                let modifies = txn.into_modifies();
+                if !modifies.is_empty() {
+                    limiter.blocking_consume(write_size);
+                    engine.modify_on_kv_engine(modifies)?;
                 }
+                Ok(())
             }
-            let write_size = txn.write_size();
-            let modifies = txn.into_modifies();
-            if !modifies.is_empty() {
-                self.limiter.blocking_consume(write_size);
-                self.engine.modify_on_kv_engine(modifies)?;
+
+            let mut keys = keys.into_iter();
+            let mut txn = new_txn(self.engine.snapshot_on_kv_engine(start_key, end_key)?);
+            let (mut next_key, mut gc_info) = (keys.next(), GcInfo::default());
+            while let Some(ref key) = next_key {
```

Comment:
Consider rename the `next_key` to `key` or something different from the `next_key` at around L208 to avoid confusion.